login/service.go

```diff
@@ -328,15 +181,222 @@ func (keycloak *KeycloakOAuthProvider) Perform(ctx *app.LoginLoginContext, confi
+		switch err.(type) {
+		case autherrors.UnauthorizedError:
+			if apiClient != "" {
+				// Return the api token
+				err = encodeToken(ctx, referrerURL, keycloakToken, apiClient)
+				if err != nil {
+					log.Error(ctx, map[string]interface{}{
+						"err": err,
+					}, "failed to encode token")
+					return nil, err
+				}
+				log.Info(ctx, map[string]interface{}{
+					"referrerURL": referrerURL.String(),
+					"api_client":  apiClient,
+				}, "return api token for unapproved user")
+				redirectTo := referrerURL.String()
+				return &redirectTo, nil
+			}
+
+			userNotApprovedRedirectURL := serviceConfig.GetNotApprovedRedirect()
+			if userNotApprovedRedirectURL != "" {
+				log.Debug(ctx, map[string]interface{}{
+					"user_not_approved_redirect_url": userNotApprovedRedirectURL,
+				}, "user not approved; redirecting to registration app")
+				return nil, autherrors.NewUnauthorizedError(err.Error())
+			}
+		}
+		return nil, err
+	}
+
+	log.Debug(ctx, map[string]interface{}{
+		"code":        code,
+		"referrerURL": referrerURL.String(),
+		"user_name":   identity.Username,
+	}, "local user created/updated")
+
+	// new user for WIT
+	if newUser {
+		err = keycloak.remoteWITService.CreateWITUser(ctx, request, identity, witURL, identity.ID.String())
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":         err,
+				"identity_id": identity.ID,
+				"username":    identity.Username,
+				"wit_url":     witURL,
+			}, "unable to create user in WIT ")
+			// let's carry on instead of erroring out
+		}
+	} else {
+		err = keycloak.updateWITUser(ctx, request, identity, witURL, identity.ID.String())
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"identity_id": identity.ID,
+				"username":    identity.Username,
+				"err":         err,
+				"wit_url":     witURL,
+			}, "unable to update user in WIT ")
+			// let's carry on instead of erroring out
+		}
+	}
+
+	err = encodeToken(ctx, referrerURL, keycloakToken, apiClient)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "failed to encode token")
+		redirectTo := referrerURL.String() + err.Error()
+		return &redirectTo, autherrors.NewInternalError(ctx, err)
+	}
+	log.Debug(ctx, map[string]interface{}{
+		"code":        code,
+		"referrerURL": referrerURL.String(),
+		"user_name":   identity.Username,
+	}, "token encoded")
+
+	if s, err := strconv.ParseBool(referrerURL.Query().Get(initiateLinkingParam)); err != nil || !s {
+		redirectTo := referrerURL.String()
+		log.Info(ctx, map[string]interface{}{
+			"referrerURL": referrerURL.String(),
+			"user_name":   identity.Username,
+			"api_client":  apiClient,
+		}, "all good; redirecting back to referrer")
+		return &redirectTo, nil
+	}
+
+	// If the 'initlinking" param == true then initiate account linking if not already linked
+	linked, err := keycloak.checkAllFederatedIdentities(ctx, keycloakToken.AccessToken, brokerEndpoint)
```

Comment:
We can now remove the all code related to linking accounts from this method. We do not use auto-linking (to github or OSO) during login anyway. Just return the redirect url without checking if the account is linked or not.

Reply:
^ Good idea :) 
@kishansagathiya you no longer need to worry about his ;)

Reply:
https://github.com/fabric8-services/fabric8-auth/pull/210/commits/ec079fc047eefd3dff08b139dd5344b90e52defe